astropy/tests/runner.py

```diff
@@ -235,23 +235,25 @@ def run_tests(self, **kwargs):
         else:
             plugins = []
 
-        # Override the config locations to not make a new directory nor use
-        # existing cache or config. Note that we need to do this here in
+        # Avoid the existing config. Note that we need to do this here in
         # addition to in conftest.py - for users running tests interactively
         # in e.g. IPython, conftest.py would get read in too late, so we need
         # to do it here - but at the same time the code here doesn't work when
         # running tests in parallel mode because this uses subprocesses which
         # don't know about the temporary config/cache.
-        astropy_config = tempfile.mkdtemp("astropy_config")
-        astropy_cache = tempfile.mkdtemp("astropy_cache")
-
-        # Have to use nested with statements for cross-Python support
-        # Note, using these context managers here is superfluous if the
-        # config_dir or cache_dir options to pytest are in use, but it's
-        # also harmless to nest the contexts
-        with set_temp_config(astropy_config, delete=True):
-            with set_temp_cache(astropy_cache, delete=True):
+        # Note, this is superfluous if the config_dir option to pytest is in use,
+        # but it's also harmless
+        orig_xdg_config = os.environ.get("XDG_CONFIG_HOME")
+        with tempfile.TemporaryDirectory("astropy_config") as astropy_config:
+            Path(astropy_config, "astropy").mkdir()
+            os.environ["XDG_CONFIG_HOME"] = astropy_config
+            try:
                 return pytest.main(args=args, plugins=plugins)
+            finally:
+                if orig_xdg_config is None:
+                    os.environ.pop("XDG_CONFIG_HOME", None)
```

Comment:
I did a double take on the second argument. I now agree this is the correct thing to do: even if this environment variable is *supposed* to be defined at this point, we cannot guarantee it is still there, and we certainly do no want to raise an exception within a `finally` clause.